File: src/statusbar/modules/datetime.ts
import type {
  StatusBarModule,
  StatusBarModuleContext,
  Segment,
} from "../types.ts";

function formatTime(format: string): string {
  const now = new Date();
  const h = now.getHours();
  const m = now.getMinutes();
  const s = now.getSeconds();
  const Y = now.getFullYear();
  const M = now.getMonth() + 1;
  const D = now.getDate();

  return format
    .replace("HH", String(h).padStart(2, "0"))
    .replace("mm", String(m).padStart(2, "0"))
    .replace("ss", String(s).padStart(2, "0"))
    .replace("YYYY", String(Y))
    .replace("MM", String(M).padStart(2, "0"))
    .replace("DD", String(D).padStart(2, "0"));
}

export const datetimeModule: StatusBarModule = {
  id: "datetime",
  render(ctx: StatusBarModuleContext): Segment[] {
    const format = (ctx.moduleConfig.format as string) || "HH:mm";
    const icon = ctx.icons ? " " : "";
    return [
      {
        text: ` ${icon}${formatTime(format)} `,
        fg: ctx.colors.fg,
        bg: ctx.colors.bg,
      },
    ];
  },
};

File: rust/crates/maxmux-statusbar/src/modules/datetime.rs
use chrono::Local;

use crate::modules::{config_str, ModuleContext, StatusBarModule};
use crate::types::Segment;

pub struct DatetimeModule;

/// Simple token-based format replacement.
/// Supported tokens: HH, mm, ss, YYYY, MM, DD.
fn format_datetime(fmt: &str) -> String {
    let now = Local::now();
    fmt.replace("HH", &format!("{:02}", now.format("%H")))
        .replace("mm", &format!("{:02}", now.format("%M")))
        .replace("ss", &format!("{:02}", now.format("%S")))
        .replace("YYYY", &format!("{}", now.format("%Y")))
        .replace("MM", &format!("{:02}", now.format("%m")))
        .replace("DD", &format!("{:02}", now.format("%d")))
}

impl StatusBarModule for DatetimeModule {
    fn id(&self) -> &str {
        "datetime"
    }

    fn render(&self, ctx: &ModuleContext) -> Vec<Segment> {
        let fmt = config_str(ctx, "format", "HH:mm");
        let formatted = format_datetime(&fmt);
        let text = if ctx.icons {
            format!(" \u{1f550} {} ", formatted)
        } else {
            format!(" {} ", formatted)
        };
        vec![Segment::new(&text, &ctx.colors.fg, &ctx.colors.bg)]
    }
}
